### ISSUE TITLE
Clarifications for libusb.info home page please

### ISSUE THREAD

**Issue by gwideman** [libusb/libusb.github.com#1]
Some issues with the current state of libusb.info.

To anyone who has googled for libusb, particularly with regard to Windows, it's evident that there are or have been multiple libusb-related projects, products and branches that either competed to do the same thing, or complement one another to create multiple components that are needed simultaneously to make things work.

I spent hours trying to figure this all out until I finally stumbled on this:
http://libusbx.1081486.n5.nabble.com/Libusbx-devel-Announcing-libusb-1-0-18-as-well-as-libusbx-1-0-18-FINAL-td2375.html

With that in mind, libusb.info could be far more helpful to clarify this confusing landscape. Focusing just on the home page, currently we have:

[1]. A logo at top left which combines "libusb" plus an "X" graphic to spell out "libusbx". This causes confusion with the libusbx project, most notably with this page: http://libusbx.sourceforge.net/, which uses the exact same logo, and is written as though the libusbx project is current and continuing and even preferable over libusb. 

Recommendation: Revise the libusb logo so that it clearly does not say "libusbx". (And obviously, http://libusbx.sourceforge.net/ should be revised to be like libusbx.org, explaining that libusbX is defunct.)

[2]. The comment "the old site is now obsolete, please use libusb.info" is completely inadequate, as it fails to specify what "old site" the new one supersedes.

Recommendation: Include some narrative that clearly describes that libusb.info is the home of THE project that supersedes both the libusb project that started at libusb.org, and also the branch that was based at libusbx.org (and reference their respective github and sourceforge homes as well, so we know these are all things that are now obsolete). Link to the email I included above as evidence.

[3]. Despite the page narrative saying that source is available on github (and indeed, libusb.info is apparently hosted by github), the Downloads menu links do a variety of suspicious-seeming things:
- Latest Source (tarball): sourceforge.net/projects/libusb/files/latest/download?source=files
- Latest Windows Binaries: ftp.heanet.ie/pub/download.sourceforge.net/pub/sourceforge/l/li/libusb/libusb-1.0/libusb-1.0.19/libusb-1.0.19-rc1-win.7z
- Other Releases> sourceforge.net/projects/libusb/files/libusb-1.0

Why on earth do the download links point to sourceforge, when github appears to be the home of this project? And why does the link for Windows binaries point to some random FTP server?

Recommendation: Either consolidate the downloadables at github and point to them there, or explain why the suspicious URLs.

[4]. There is apparently also a libusbK project. Please mention this, and how it relates to libusb, if at all.

[5]. There's WinUSB. Again, please note this, and how it relates to libusb.

[6]. Please mention libusb-win32 and whether it is a subsidiary project, a sister project, or a completely independent project that either does or does not have some interplay with libusb or the libusb project.

I hope these suggestions are easy to complete. If done, they will make life much easier for all who stumble in here trying to sort out USB confusion.

Thanks.

**Comment by pbatard** [comment]
> or complement one another to create multiple components that are needed simultaneously to make things work.

No, that's wrong. All you need to use libusb on Windows is the libusb library (from libusb.info) and the Microsoft WinUSB driver. You don't need to care about libusb-win32, libusbx or libusbK unless you want to.
The Windows **backend** page (the backend page is NOT the main page) tries to clarify that, but probably not as good as it can, because it's hard for people close to a project to figure out the points of contentions newcomers might have.

[1] logo issue

The 'X' is derived from the standard USB logo and, in its prime **artistic** interpretation, in the context of libusb, should be seen as a cross for 'cross platform', which, coincidentally, is part of the tagline that comes right after the logo. A conscious decision was kept to keep the libusbx logo, because it actually works for both, and libusbx is in the process of being phased out, therefore the only confusion people have, which happens if they saw the libusbx, will be alleviated. But libusbx is being phased out, and will effectively disappear from the internet once the libusbx.org domain expires. Then again, please remember, a logo is just a logo. If you try to infer information from it, you're not looking at it for what it is.

[2] Conscious decision there again in order to to advertise for the .org site, because we want the .info to be returned first by search engines when looking for libusb (and we're still a long way from reaching this goal), precisely to AVOID the confusion from people ending up with a library/project (the one from the .org) that hasn't been updated for years.

[3] All the page says the **source** is available on github. There's no contract or de-facto rule that says downloads must be available from github too, and as a matter of fact, when we started with libusb.info, github had just taken active steps to remove easy download facilities from their site (which they have somewhat backtracked upon, but it took them months to do that). Combined with the fact that SF is the place where people coming from the .org site will be fetching libusb tarballs and binaries (though an older version of them), keeping the libusb.info downloads there gives a chance for these users to realize that there exist newer versions and, hopefully, download them. If we were hosting our downloads elsewhere, the .org users would never had a chance to see these updated versions of the libusb library, and miss out on critical bugfixes.

As to the Windows binaries, the explanation is that SF has taken a somewhat hostile stance to direct downloads (need to wait 5 secs with unwanted advertising), so we use an **official** FTP SF mirror to avoid the 5 secs delay and the ads, to provide a better user experience on this specific download.

[4] I think mentioning libusbK and libusb-win32, when they are:
- irrelevant for non Windows users (remember libusb is cross platform and also support Linux, OS X, Android, *BSD and Haiku - people using these platforms don't care about Windows specifics)
- not needed to use libusb

Will be more confusing than anything else. As a matter of fact, it looks to me like you are the one who got confused by the mention of libusbK and libusb-win32 from the libusb Windows backend page, which you don't really need to care or know about to be able to use libusb on Windows (see next point).

[5] To use libusb on Windows, you need an USB driver (app &rarr; libusb library &rarr; [libusb-compatible driver] &rarr; USB device). Normally, [libusb-compatible-driver] is the WinUSB driver from Microsoft, but, if you want, you can also use the **driver part** from the libusb-win32 and libusbK projects (that, unlike libusb, which is only a library, are projects that provide both a library and a driver). I think that's the information you were missing from the start. 

Now, since you have a better grasp than us on what potentially confusing information newcomers might want to see clarified on the wiki pages, and the wiki is opened to everybody (AFAIK - you just need to register a github account), instead of asking us to modify it, and most likely fail to address your concerns, how about you do that yourself? It's usually a lot better for the person who struggled with understanding something to clarify it for the next person, and anyone with a github should be able to edit https://github.com/libusb/libusb/wiki/Windows.

[6] see [4]

Last but not least, you do have a very valid point about libusbx.sourceforge.net. But it just happens that we forgot to delete it when we moved to libusbx project homepage to http://libusbx.org/ (which mentions that libusbx is now obsolete and should be ignored altogether). I'll take steps to remove that page altogether.

**Comment by pbatard** [comment]
The index page from http://libusbx.sourceforge.net/ has now been removed (and now point to SF default content that does mention the project to be obsolete and superseded by libusb.info).

**Comment by gwideman** [comment]
@pbatard Thanks for your lengthy and detailed response -- very helpful.

Thanks for taking steps to remove leftover libusbx sites that were a distraction, that definitely clarifies the landscape. 

(I dispute your assertion that the "X" part of the libusbx (and now libusb) logo would be interpreted by readers as a cross for cross-platform, when this logo previously represented libusbx, for which it only made sense by reading the "libusb" text combined with the "X" graphic. But that's water under the bridge -- if there's no distracting libusbx project vying for attention, problem solved.)

I would like to contribute to the wiki as you suggest, on page https://github.com/libusb/libusb/wiki/Windows.  However, could you first spell out what this existing sentence means?

"This project adds Windows platform support to the libusb Open Source library".

I am confused because I am assuming that "this project" refers to libusb, but then the sentence goes on to say that "this project" adds  Windows support to libusb.

So is there some distinction between "this project" and libusb?

-- Graham

**Comment by LudovicRousseau** [comment]
I agree with @gwideman that using a "strange" ftp site for Windows binary is disturbing.
I do not use Windows so I had not noticed this URL before.

The goal of sourceforge mirrors is to provide the "best" server for our users. I don't think the libusb project should play with sourceforge mirror system.

AdBlock+ is efficient to remove the ads. I only get the 5 seconds delay :-)

**Comment by gwideman** [comment]
@LudovicRousseau  FWIW, @pbatard has convinced me that the ftp site is not actually sinister, and I now understand the annoying sourceforge behavior that libusb.info is trying to work around. It all makes sense when you know the background.

However, I can't help feeling that sticking with sourceforge (or github) as the file host is a better idea from the standpoint of credibility.  Given the disarray in the libusb landscape (hopefully to be resolved), people stumbling in here don't really know which libusb project is real, and which might be a phishing or trojan site, and the ftp site doesn't help matters. Especially because it's particular to the binaries -- executables that need to be installed on faith that they are not malicious.

**Comment by pbatard** [comment]
@LudovicRousseau 

> AdBlock+ is efficient to remove the ads. I only get the 5 seconds delay :-)

Wow, you're actually happy with the delay? This (and the overall direction of the DICE network of late) is the part I'm actually riled up against. [_Cue in another of Pete's rants_] If it was just the ads, and the download was instantaneous I wouldn't care much about linking to the official SF link for the Windows binaries (the tarball is a different story mostly because we don't need to update that link for each release if we do things well - but I can't say I haven't itched to do the same for the tarball). But forcing the world to collectively waste millions of seconds is unforgivable. For instance, let's say 1M in the current [estimated 5M monthly visitors of SF](https://siteanalytics.compete.com/sourceforge.net/) comes there for a download: you end up with that stupid wait policy collectively forcing the whole planet to waste more than 50 days in forced delays every month (or more than 1.5 day **every day**). How'd you like if I delayed all your downloads by 1.5 or 50 days for no good reason. But hey, this time wastage is not localized to a single individual, so it's not the same thing at all, right? Who gives a damn about what happens to the world collectively, even when we have very efficient means of evaluating the actual damage...

Also, considering that I'm using heanet on a very regular basis, have confirmed that they can provide very decent speed no matter where you are located in the world (I've used it a few times outside of EU), and have seen more outages of SF than heanet (latest SF outage was a few days ago), I posit that we are actually better off using direct links off what I have to stress out again is a **very official and verifiable** SF mirror than SF itself. But feel free to update the download link if you feel that strongly about it, coz I'm not gonna bother. Furthermore, I also feel that whoever provides a download archive should have the right to decide where they choose to put it, even if it collides with some people's ill placed views of what a "safe" download should look like - heck I might very well have decided to host them on akeo.ie to get the kind of detailed download statistics I get from some of my [other](http://rufus.akeo.ie) [open source](http://zadig.akeo.ie) [offerings](http://efi.akeo.ie), for which I publish the source on github but use a completely different download site, in part to get a better idea of who downloads what.

@gwideman 

First, you'd be surprised about the amount of thought that went into the logo, down to the colour choice. I was the one who designed it, and yes, when I did that, I did consider that we might reuse it for libusb if we ever had a chance (though at the time, I thought it'd make sense to move the stylized 'X' part to the front of libusb, but after seeing the logo in situ, I no longer see it as required).

> "This project adds Windows platform support to the libusb Open Source library"

This is just a historical leftover from the many months during which the Windows backend of libusb had to exist as a separate project/branch due to its constantly delayed integration into the project. As you already found out, keeping documentation up to date and relevant is a harder and more time-consuming task than it looks

I guess a more proper current wording would be "This page details the specifics of the Windows backend part of libusb, which helps developers...". It may also make sense to detail what a "backend" is, which in layman's terms could probably be defined as "the part of a cross-platform project that is specific to an OS".

At any rate, your willingness to help **is** appreciated!

EDIT: daily &rarr; monthly

**Comment by gwideman** [comment]
@pbatard Hi Pete: First let me say that this discussion is against a background in which I, for one, greatly appreciate the work that goes in to open source efforts such as this one, and the tremendous amount of industry activity (and livelihoods) which is facilitated by crucial components such as this. Huge credit to you and other participants for your labors and inspirations.

So it's only natural that an audience member such as myself would want your efforts to gain as much traction as possible. I don't dispute at all your right to decide where to host a download, absolutely of course that's up to you.  I am simply reporting that, to the visitor and potential user, neither the authoritativeness of libusb.info, nor your reasons for hosting downloads on a "mystery ftp site", are readily apparent (though improved with the recent muting of libusbx offerings). This is somewhat tragic, as it dilutes the tremendous work you have done. And, I suggest, for viewers it precipitates at least five seconds of confusion and hesitation, outweighing SF's five second delay.

But -- no need to compromise. It's my belief that libusb.info could easily describe a bit more of the context of the libusb landscape, and thereby cultivate visitors' confidence that indeed this is the authoritative site. And the "anomalous" download URLs could be explained in one sentence to alleviate any concerns there (at the same time airing your critique of SF's download delay and ginning up some commiseration).  Also, you mention that the download site is a verifiable mirror -- I'm interested to know how does one verify that?

Wiki: Thanks for the explanation. I will think through what I might contribute. I'd like to explain (and diagram), at least for Windows, how the various mentioned products and components fit into the scenario of a Windows application talking to some external equipment. Since I don't actually know how that works, I've started poking around in the MSDN docs, and looking at Russinovich's Windows Internals Part 2, trying to make sense of it. Feel free to suggest other sources (especially with diagrams) that might be useful.

**Comment by pbatard** [comment]
1. The homepage is designed to be short and to the point. It is really not the place to go into a tirade against the direction SF has taken of late or other matters that aren't directly relevant to the needs of software developers wanting to use a library. I'm confident that 99.9% of visitors won't give a damn about that or where the site owners decided to link to when it comes to downloads and if I were to do extra work to address the request of what will remain a **minority of users**, I'd start with ensuring that the site is properly accessible for vision impaired users or something. Also, as I explained, the "anomalous" download is not relevant for non Windows users, or people who aren't going to download Windows binaries. Moreover, if the site was really malicious, or had been compromised, it would be exceedingly easy to link to compromised Windows binaries behind an URL that _looks_ more legit, including a project such as "libusb_downloads" that would have been created on SF for malicious purpose for instance.  
   Also I dispute your claim that this will incur >5s confusion on the vast majority of people. Most people will see a link placed by the site owners, and trust that the people who created the software they came here to use are linking to a legitimate download. If you can't do that, then you are assuming that libusb.info as a whole is a site for rogue software, which you probably shouldn't visit in the first place.  
   Finally, if you really don't trust the link, you can go to the libusb repository on sourceforge **yourself** and download the Windows binaries from there. So even if you are paranoid about provenance, you have the means to do something about it.
2. If you clear your browser cache and go download anything from SF, you'll be presented with a choice of mirrors. heanet is one of those choices. (NB: you need to clear your browser cache, or at least the entries related to SF if you already downloaded anything, as your browser will then remembed your mirror preference).
3. Wiki: If you plan to upload pictures, we're probably going to have to do that for you, as this needs to be done through git, and only the project owners can push changes with git. AFAIK, github has no easy way to add pictures with the standard web interface (that is unless you were to upload the picture to an external site).
